tryWinKernelspaceProxy: handle kernelspace not allowed, make ServiceInfo same as kernelspace, comment all functionality WRT go proxy

jayunit100 opened the pull request:
**What type of PR is this?**

<!--
/kind cleanup
-->

**What this PR does / why we need it**:

Make it easier to reason about the kube-proxy windows codepath

**Which issue(s) this PR fixes**:

No issue, this is just a cleanup PR mostly 


**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

jayunit100 committed:
Cleanup of kube-proxy windows, decoupling of trying,
userspace/kernelspace, logging for newly created service proxies
(up/down), explaining netsh usage, and comments for all functions.

JocelynBerrendonner reviewed (commented):
LGTM from a Windows perspective. That said, let me include a few additional folks to handle Windows testing and signoffs from a Windows perspective.

JocelynBerrendonner commented:
@sbangari : Can I let you handle the signoff for this PR from a Windows perspective, please?

jayunit100 commented:
Thanks, if we're all on board I'll clean this Pr up tomorrow

sbangari commented:
/lgtm

jayunit100 commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

jayunit100 commented:
/retest

fedebongio commented:
/remove-sig api-machinery

jayunit100 commented:
/retest

jayunit100 commented:
@sbangari needs another review, i rebased and cleaned up the stray ends.

jayunit100 commented:
Ok, i got it up and running on a windows machine to test .  Seems to work fine for me

```
    InterfaceMetric:                      15

I1120 03:20:11.632652   13320 proxier.go:245] Proxying for service "default/kubernetes:https:100.2.2.1" at 100.2.2.1:443/TCP
I1120 03:20:13.735757   13320 proxysocket.go:179] Accepted TCP connection from 100.2.2.246:53694 to 100.2.2.246:443
I1120 03:20:13.735757   13320 proxysocket.go:139] Mapped service "kube-system/antrea::100.2.2.246" to endpoint 10.0.0.42:10349
I1120 03:20:13.737876   13320 proxysocket.go:179] Accepted TCP connection from 100.2.2.246:53696 to 100.2.2.246:443
I1120 03:20:13.737876   13320 proxysocket.go:139] Mapped service "kube-system/antrea::100.2.2.246" to endpoint 10.0.0.42:10349
I1120 03:20:13.740741   13320 proxysocket.go:179] Accepted TCP connection from 100.2.2.246:53698 to 100.2.2.246:443
I1120 03:20:13.740741   13320 proxysocket.go:139] Mapped service "kube-system/antrea::100.2.2.246" to endpoint 10.0.0.42:10349
I1120 03:20:19.577651   13320 proxysocket.go:179] Accepted TCP connection from 100.2.2.1:53700 to 100.2.2.1:443
I1120 03:20:19.577651   13320 proxysocket.go:139] Mapped service "default/kubernetes:https:100.2.2.1" to endpoint 10.0.0.42:6443

```

jayunit100 commented:
/retest

jayunit100 commented:
@JocelynBerrendonner can you give this another LGTM ? i made some trivial changes ...  who should be the secondary approver ?

jayunit100 commented:
@sbangari can you look one more time? cleaned up a couple of other godoc functions

sbangari commented:
/lgtm for windows changes

sbangari commented:
/lgtm

JocelynBerrendonner commented:
/lgtm

andrewsykim commented on cmd/kube-proxy/app/server_windows.go:
```diff
@@ -178,31 +178,33 @@ func newProxyServer(config *proxyconfigapi.KubeProxyConfiguration, cleanupAndExi
+		}
+		return proxyModeActual
+	} else {
+		klog.V(1).Infof("Using userspace proxy, unknown proxy input %v", proxyMode)
```
"Unknown proxy mode %q, using userspace proxy"

jayunit100 replied:
+1 will do that shortly

andrewsykim commented on cmd/kube-proxy/app/server_windows.go:
```diff
@@ -178,31 +178,33 @@ func newProxyServer(config *proxyconfigapi.KubeProxyConfiguration, cleanupAndExi
-		return tryWinKernelSpaceProxy(kcompat)
+		proxyModeActual, err := tryWinKernelSpaceProxy(kcompat)
+		if err != nil {
+			fmt.Errorf("Can't determine whether to use windows kernel proxy %v (using proxy mode: %v)", err, proxyModeActual)
```
This is just instnatiating an error but not doing anything with it. Was this supposed to be a `klog.Error` or `klog.Warning?

jayunit100 replied:
OOPS ! :) i gotta fix this one

jayunit100 replied:
Fxd

andrewsykim commented on cmd/kube-proxy/app/server_windows.go:
```diff
@@ -178,31 +178,33 @@ func newProxyServer(config *proxyconfigapi.KubeProxyConfiguration, cleanupAndExi
+			fmt.Errorf("Can't determine whether to use windows kernel proxy %v (using proxy mode: %v)", err, proxyModeActual)
+		}
+		return proxyModeActual
+	} else {
```
I think we can remove this else

jayunit100 replied:
Done

andrewsykim commented on pkg/proxy/config/config.go:
```diff
@@ -107,6 +107,8 @@ type EndpointsConfig struct {
 
 // NewEndpointsConfig creates a new EndpointsConfig.
 func NewEndpointsConfig(endpointsInformer coreinformers.EndpointsInformer, resyncPeriod time.Duration) *EndpointsConfig {
+	klog.Infof("creating new endpoints config with resync period of %v", resyncPeriod)
```
Log lines usually start with capital letter. Also wondering why this log message is needed, is it to easily determine if we're watching Endpoints or EndpointSlice?

jayunit100 replied:
The reason i actually added this was because of the generally confusing state of affairs when it comes to kube-proxy config.  We dont print out the entire configuration, and so if someone plumbs a custom value in, its super helpful to know what that value was :)

andrewsykim replied:
This isn't actually "config" though, this is the endpoints watcher/handler logic named EndpointsConfig.

jayunit100 replied:
hmmm  are you suggesting we shouldnt log it , or that its incorrect to call it a config, or both ?   resyncPeriod is definetly a configuration parameter (  https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-proxy/app/server.go#L201 ) at least, in the sense that it is something that varies and is sent in by the user and configured from a configMap ?

andrewsykim replied:
resyncPeriod is a configuration option passed into the EndpointsConfig, but EndpointsConfig does not actually parse or read the configuration for it, that is done in a previous step by kube-proxy when CLI args are read. 

I do think the name `EndpointsConfig` is a misleading name but I don't think it's worthwhile changing in this PR.

jayunit100 replied:
ok, so @andrewsykim all i really care about is that we convey the kube-proxy input arguments **somehow** .  this is one of many such args.  Maybe we can do that in a follow on PR somehow ? i.e. just dump all configuration params that are being used on startup ?

andrewsykim replied:
Other components do dump all flags set on start up, but I think it's only at a certain verbosity level. Recommend checking what kubelet does and do similar? Would prefer a follow-up PR for that

jayunit100 replied:
I think this is best done via the keep above so... just to expedite this I’ll  move this effort there.....

andrewsykim commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -1117,13 +1133,6 @@ func iptablesFlush(ipt iptables.Interface) error {
 	return utilerrors.NewAggregate(el)
 }
 
-// Used below.
```
What was the reason for moving this to the top of the file?

jayunit100 replied:
they are "basically" constants, normally those would go at the top of a file. maybe would be better if i just consolidated these somewhere ?  its not like `127.0.0.1` needs a bespoke implementation for all 4 proxy types :) ?

andrewsykim commented on pkg/proxy/userspace/roundrobin.go:
```diff
@@ -151,6 +151,7 @@ func (lb *LoadBalancerRR) NextEndpoint(svcPort proxy.ServicePortName, srcAddr ne
 	sessionAffinityEnabled := isSessionAffinity(&state.affinity)
 
 	var ipaddr string
+
```
why this extra line here?

jayunit100 replied:
will fix

andrewsykim commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -106,8 +106,8 @@ type loadBalancerFlags struct {
-// internal struct for string service information
-type serviceInfo struct {
+// ServiceInfo is an internal struct for string service information
+type ServiceInfo struct {
```
What's the reason for exporting this struct? It seems specific to winkernel so should remain private

jayunit100 replied:
ah yes, see comment below just dug it up

andrewsykim commented on pkg/proxy/winuserspace/proxier.go:
```diff
@@ -39,13 +39,17 @@ import (
 
-type serviceInfo struct {
+// ServiceInfo contains information and state for a particular proxied service
+type ServiceInfo struct {
```
same question here, what's the reason for exporting this?

jayunit100 replied:
iirc the other `ServiceInfo` instances were doing this, but i need to double check... will get back on this one

jayunit100 replied:
ah yes found it here https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/userspace/proxier.go
 ...  Im assuming here that the Linux kube-proxy is the gold standard... and this is a deviation from that:

```
// ServiceInfo contains information and state for a particular proxied service
54 | type ServiceInfo struct {
55 |  
```

is what we do in pkg/proxy/userspace/proxier.go. Somehow that comment, and the struct exporting, both changed when copied to windows .... its a minor difference but, when comparing the two implementations...  it makes you wonder why one is not exported, while the other **is** exported.

andrewsykim commented on pkg/proxy/winuserspace/proxysocket.go:
```diff
@@ -510,10 +512,9 @@ func (udp *udpProxySocket) ProxyLoop(service ServicePortPortalName, myInfo *serv
 		}
 	}
-
+	klog.V(3).Infof("ProxyLoop is now starting for service %v", myInfo)
```
This seems like it could be pretty noisy, even for v=3

jayunit100 replied:
... hmmmm...  here was my thinking - but i wrote this PR like a month ago so i mighta forgot something --- lmk what im missing here -- 

- The goal is to make the userspace kube proxy logs `literate`, i.e. to tell the story around whats happening by default, given the complexity of managing services in user space .....    

- the default logging would not be modified by this - as you implied... so  at `v=3`,  you get  a new  log statement anyways  `for every new  service` iirc already which occurs in kube proxy .... So we're not algorithmically increasingly the amount of noise.  Also remember this code base is much more dangerous and difficult to reason about then the hardened linux kube proxy, so it could save us lots of time in production scenarios 

- also remember, every cni has bespoke kube proxy instructions, so this noise is less trivial then you might initially think?

andrewsykim commented on pkg/proxy/winuserspace/roundrobin.go:
```diff
@@ -53,8 +53,8 @@ type affinityPolicy struct {
-	lock     sync.RWMutex
-	services map[proxy.ServicePortName]*balancerState
+	lock                    sync.RWMutex
+	winLoadbalancedServices map[proxy.ServicePortName]*balancerState
```
I think old field name `services` is fine, "loadbalanced" is implied since the type is named `LoadBalancerRR`

jayunit100 replied:
ok fair - i can move that back to what it was

jayunit100 replied:
Done resolvinggg

jayunit100 commented on pkg/proxy/winuserspace/proxier.go:
```diff
@@ -253,7 +262,8 @@ func (proxier *Proxier) addServicePortPortal(servicePortPortalName ServicePortPo
 }
 
-func (proxier *Proxier) closeServicePortPortal(servicePortPortalName ServicePortPortalName, info *serviceInfo) error {
+// closeServicePortPortal
```
self note i need to finish commenting this

jayunit100 committed:
reset services

jayunit100 committed:
update closeServicePortPortal

jayunit100 committed:
remove logging

jayunit100 committed:
review stuff

jayunit100 committed:
review stuff

jayunit100 committed:
review stuff

jayunit100 commented:
Addressed all comments left responses to other ones.... 
- “constants” ... should we move them to a single file ? Or is top of file ok ?
- ServiceInfo seems to be the norm for other places.. can change it back but seems like it really helps me when comparing proxy implementations for them to follow the same naming conventions 
- regarding proxy service log levels ... could debate it either way but ... V3 is logging new service stuff already so, I don’t think we’re adding “algorithmically” adding a new noise category ... but if it’s a hard requirement I’m happy to move that log to v4... keep in mind for the royal  “us” (me :)) the motivation here is explicit easy to debug behavior , and not many people use this code path which is why .....  logging verbosity is relatively important compared w  other places  

Anyways flexible on these 3 so just lmk.. final thoughts on those thanks for the review !

jayunit100 committed:
Add compilation deets

ehashman commented:
/sig windows

needs rebase

jayunit100 commented:
i think im going to close this and do it peice meal
